### DISCUSSION TITLE
fortls initialization never ending within Visual Studio Code - Modern fortran

### DISCUSSION THREAD

**Discussion by cgattano** [fortran-lang/fortls#349] [Q&A]
Dear community,

I would like to report a situation that I have faced in the institute in which I worked regarding fortls and Visual Studio code -- modern fortran. **The problem is that fortls starts via modern fortran extension to initialize itself but never ends** (I mean there is no usual error message showing up that other users may have reported in the other discussions available). After approximately one hour, our server started to highly lag because the fortls jobs was still running in background and took all the memory available on our server.

I am not sure what information would be required to give you the possibility to provide helps or advises on how to investigate this issue (this is the first time I write a request message on a forum so I may not know the code of a good request message), so feel free to ask me for additional information. I just provide details that I think may be interesting in the discussion:
* We work with VS code locally installed on our own computer and ssh-connect to our Linux-based server where the source files are stored.
* **fortls version** used 2.13.0
* **modern fortran version** used v3.4.2024011701
* Our software is composed of a very large mixture of all types of fortran files (from the very old-time standard to fortran 2003 standards [and a lot of non-standard things...]) started more than 30 years ago. (I guess you understand the picture ;)...)
* The problem happens with apparently one specific version of our software. For most of our version it works properly and fortls initialization reaches a successful end.

This is a screenshot of the modern fortran output with the last messages, after that no more output and fortls job runs and never ends
![fortls_initializing](https://github.com/fortran-lang/fortls/assets/104211095/975e72c6-91b3-4066-8001-7c50829ac4d0)

We would like to be able to investigate what could have happened with fortls around this specific version of our software, but we missed some ideas on where and how to start investigation. We are scientists and are a bit limited in our informatics abilities ^^'.

Would someone have some advises on what he would do to start investigating the problem ?

Thanks in advance for your ideas and support,
Best regards,

César

**Comment by gnikit** [comment]
It sounds like the server is getting stuck somewhere or simply there are too many files so it takes ages to parse them.
The hardware that you have in the server matters (free CPU cores and free RAM). Every instance of Modern Fortran in VS Code will spin up its own version of `fortls`. By default, each `fortls` processes will use 4 cores when initialising. You can change that number from the options https://fortls.fortran-lang.org/options.html.

The best way to get to the bottom of this is to share the version of the codebase that causes the issue. If that is not an option I can try and walk you through debugging the issue using the `fortls` command line interface, see: https://fortls.fortran-lang.org/options.html#debug-options-command-line-only. However, my availability over the next few months is very limited so 1-on-1 open-source tech support is unlikely.

**Comment by cgattano** [comment]
Thanks for your quick answer gnikit,

I indeed am not allowed to share source code, unfortunately. And identifying the file in question that creates the problem will take us some time because of the number of files involved (~100 programs and ~1500 modules). For comparison, for a version of the project with which fortls is initializing correctly, it takes 10 to 30 sec to finalize initialization.

Passing the number of core used from 4 to 1 would be already a way to minimize the impact of the issue when it happen until we understand what happen, thanks for pointing me out this option.

I understand and would not wish to bother you too much with that. The minimum advise you can provide is already highly welcome.
I will try to play a little bit with the debugging options. Shall I used the fortran > fortls: extra args setting to play with the different debug options within modern fortran or would you recommend me to leave the modern fortran environment at the moment and run fortls with these options directly in a terminal ?

**Comment by gnikit** [comment]
My advice would be to;

1. ssh into the server from a terminal (don't launch vscode)
2. run `pip install -U fortls --pre` to install the latest pre-release
3. cd into your project directory
4. run `fortls --debug_rootpath .`, this is the simplest way to initialise the server and check if something goes wrong
5. Experiment with the 
   1. `-n NUM_THREADS` options, e.g. set to 1 and see if something changes
   2. `--debug_parser --debug_filepath SOME_SOURCE_FILE`, to check if a file is parsed correctly. Start working the files that carry the most diffs from a working version of your codebase to the one that is causing the breaks.

I hope this helps, feel free to update this discussion with any news since you might have stumbled onto a bug.
